File: vyoma-ui/src/lib/subscriptionAccess.ts
import type { VyomaUserProfile } from '../types/userProfile'

export function hasActiveSubscription(profile: VyomaUserProfile | null): boolean {
  if (!profile) return false
  if (profile.subscriptionStatus !== 'active') return false
  if (!profile.subscriptionExpiresAt) return false
  return profile.subscriptionExpiresAt.getTime() > Date.now()
}

export function subscriptionInactiveReason(profile: VyomaUserProfile | null): string {
  if (!profile) return 'Sign in and choose a plan or redeem a coupon.'
  if (profile.subscriptionStatus !== 'active') {
    return 'No active subscription. Subscribe or redeem your 1-month coupon.'
  }
  if (!profile.subscriptionExpiresAt) return 'Subscription data missing.'
  if (profile.subscriptionExpiresAt.getTime() <= Date.now()) {
    return 'Your subscription expired. Renew at vyomai.app.'
  }
  return ''
}

File: lib/core/subscription_access.dart
import 'models/user_profile.dart';

/// Server-mirrored entitlement check (Firestore `users/{uid}`).
abstract final class SubscriptionAccess {
  static bool hasActive(UserProfile? profile) {
    if (profile == null) return false;
    if (profile.subscriptionStatus != 'active') return false;
    final expires = profile.subscriptionExpiresAt;
    if (expires == null) return false;
    return expires.isAfter(DateTime.now());
  }

  static String? inactiveReason(UserProfile? profile) {
    if (profile == null) return 'Sign in and choose a plan or redeem a coupon.';
    if (profile.subscriptionStatus != 'active') {
      return 'No active subscription. Subscribe or redeem a 1-month coupon.';
    }
    final expires = profile.subscriptionExpiresAt;
    if (expires == null) return 'Subscription expiry missing. Contact support.';
    if (!expires.isAfter(DateTime.now())) {
      return 'Your subscription expired. Renew at vyomai.app to continue.';
    }
    return null;
  }
}
